File: loopx/control_plane/work_items/delivery_outcome.py
from __future__ import annotations

from collections.abc import Mapping
from enum import Enum
from typing import Any

from .progress_result import (
    PROGRESS_OBSERVATION_SCHEMA_VERSION,
    ProgressResultClass,
    normalize_progress_identifier,
)


class DeliveryOutcome(str, Enum):
    """Structured machine signal for what a delivery run actually advanced."""

    SURFACE_ONLY = "surface_only"
    OUTCOME_GAP = "outcome_gap"
    OUTCOME_PROGRESS = "outcome_progress"
    PRIMARY_GOAL_OUTCOME = "primary_goal_outcome"


class DeliveryTurnKind(str, Enum):
    """Compact public-safe classification for why a delivery turn counts."""

    CONTRACT_ONLY_PREPARATION = "contract_only_preparation"
    COMPACT_EVIDENCE = "compact_evidence"
    BLOCKER_WRITEBACK = "blocker_writeback"
    PRODUCT_PATH_EXECUTION = "product_path_execution"
    OUTCOME_GAP = "outcome_gap"
    UNKNOWN = "unknown"


DELIVERY_OUTCOME_CHOICES = tuple(outcome.value for outcome in DeliveryOutcome)
DELIVERY_TURN_KIND_CHOICES = tuple(kind.value for kind in DeliveryTurnKind)

MATERIAL_DELIVERY_OUTCOMES = frozenset(
    {
        DeliveryOutcome.OUTCOME_GAP,
        DeliveryOutcome.OUTCOME_PROGRESS,
        DeliveryOutcome.PRIMARY_GOAL_OUTCOME,
    }
)
ACCOUNTABLE_DELIVERY_OUTCOMES = frozenset(
    {
        DeliveryOutcome.OUTCOME_PROGRESS,
        DeliveryOutcome.PRIMARY_GOAL_OUTCOME,
    }
)
PROGRESS_DELIVERY_OUTCOMES = ACCOUNTABLE_DELIVERY_OUTCOMES


def qualifies_turn_scoped_blocker_settlement(
    delivery_outcome: Any,
    progress_observation: Mapping[str, Any] | None,
    *,
    work_item_id: str | None = None,
    replan_obligation_id: str | None = None,
) -> bool:
    """Return whether an outcome gap is a typed, attributable blocker receipt.

    ``outcome_gap`` remains outside the progress outcomes: it can settle one
    exact Turn only when the same writeback carries a blocked observation with
    a stable blocker, evidence, and matching Todo or replan-obligation identity.
    """

    if (
        normalize_delivery_outcome(delivery_outcome) != DeliveryOutcome.OUTCOME_GAP
        or not isinstance(progress_observation, Mapping)
        or progress_observation.get("schema_version")
        != PROGRESS_OBSERVATION_SCHEMA_VERSION
        or not isinstance(progress_observation.get("evidence_ids"), list)
        or bool(work_item_id) == bool(replan_obligation_id)
    ):
        return False
    observation = dict(progress_observation)
    normalized_work_item_id = normalize_progress_identifier(
        work_item_id or replan_obligation_id
    )
    if (
        normalized_work_item_id is None
        or observation.get("result_class") != ProgressResultClass.BLOCKED.value
        or normalize_progress_identifier(observation.get("blocker_id")) is None
        or normalize_progress_identifier(observation.get("work_item_id"))
        != normalized_work_item_id
    ):
        return False
    evidence_ids = observation.get("evidence_ids")
    return isinstance(evidence_ids, list) and bool(evidence_ids) and all(
        normalize_progress_identifier(evidence_id) is not None
        for evidence_id in evidence_ids
    )


def qualifies_turn_scoped_settlement(
    delivery_outcome: Any,
    progress_observation: Mapping[str, Any] | None,
    *,
    work_item_id: str | None = None,
    replan_obligation_id: str | None = None,
) -> bool:
    """Return whether one delivery record may satisfy a Turn settlement."""

    normalized = normalize_delivery_outcome(delivery_outcome)
    return normalized in ACCOUNTABLE_DELIVERY_OUTCOMES or (
        qualifies_turn_scoped_blocker_settlement(
            normalized,
            progress_observation,
            work_item_id=work_item_id,
            replan_obligation_id=replan_obligation_id,
        )
    )


def normalize_delivery_outcome(value: Any) -> DeliveryOutcome | None:
    if isinstance(value, DeliveryOutcome):
        return value
    text = str(value or "").strip()
    if not text:
        return None
    try:
        return DeliveryOutcome(text)
    except ValueError:
        return None


def require_delivery_outcome(value: Any) -> DeliveryOutcome:
    outcome = normalize_delivery_outcome(value)
    if outcome is None:
        raise ValueError("delivery_outcome must be one of: " + ", ".join(DELIVERY_OUTCOME_CHOICES))
    return outcome


def delivery_outcome_value(value: Any) -> str | None:
    outcome = normalize_delivery_outcome(value)
    return outcome.value if outcome else None


def normalize_delivery_turn_kind(value: Any) -> DeliveryTurnKind | None:
    if isinstance(value, DeliveryTurnKind):
        return value
    text = str(value or "").strip()
    if not text:
        return None
    try:
        return DeliveryTurnKind(text)
    except ValueError:
        return None


def require_delivery_turn_kind(value: Any) -> DeliveryTurnKind:
    kind = normalize_delivery_turn_kind(value)
    if kind is None:
        raise ValueError("delivery_turn_kind must be one of: " + ", ".join(DELIVERY_TURN_KIND_CHOICES))
    return kind

File: loopx/control_plane/work_items/delivery_outcome.ts
import { EffectRuntimeRequestError } from "../effect_runtime_errors.ts";
import { requireJsonObject } from "../runtime_decode.ts";
import type { JsonObject } from "../effect_program.ts";

export const DELIVERY_OUTCOMES = [
  "surface_only",
  "outcome_gap",
  "outcome_progress",
  "primary_goal_outcome",
] as const;

export const MATERIAL_DELIVERY_OUTCOMES = [
  "outcome_gap",
  "outcome_progress",
  "primary_goal_outcome",
] as const;

export type DeliveryOutcome = (typeof DELIVERY_OUTCOMES)[number];
export type MaterialDeliveryOutcome =
  (typeof MATERIAL_DELIVERY_OUTCOMES)[number];

export type DeliveryClaimConflict =
  | "progress_with_preparation_only"
  | "primary_outcome_with_blocker"
  | "primary_outcome_with_followthrough";

/** Validate combinations, not prose or evidence truth. Historical readers use
 * these same codes without rewriting already committed records or receipts. */
export function diagnoseDeliveryClaim(raw: JsonObject): DeliveryClaimConflict[] {
  const outcome = String(raw.delivery_outcome ?? "").trim();
  const kind = String(raw.delivery_turn_kind ?? "").trim();
  const conflicts: DeliveryClaimConflict[] = [];
  if ((outcome === "outcome_progress" || outcome === "primary_goal_outcome")
    && kind === "contract_only_preparation") conflicts.push("progress_with_preparation_only");
  const observation = jsonObject(raw.progress_observation);
  if (outcome === "primary_goal_outcome" && (kind === "blocker_writeback"
    || (observation?.schema_version === "typed_progress_observation_v0" && observation.result_class === "blocked"))) {
    conflicts.push("primary_outcome_with_blocker");
  }
  if (outcome === "primary_goal_outcome" && raw.outcome_followthrough_required === true) {
    conflicts.push("primary_outcome_with_followthrough");
  }
  return conflicts;
}

export function validateDeliveryClaim(value: unknown): JsonObject {
  const conflicts = diagnoseDeliveryClaim(requireJsonObject(value, "delivery claim"));
  return { schema_version: "delivery_claim_validation_v0", valid: conflicts.length === 0, conflicts };
}

const STABLE_ID_PATTERN = /^[A-Za-z0-9][A-Za-z0-9._:/-]{0,127}$/;
const PROGRESS_DELIVERY_OUTCOMES = new Set<DeliveryOutcome>([
  "outcome_progress",
  "primary_goal_outcome",
]);

function jsonObject(value: unknown): Record<string, unknown> | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function stableIdentifier(value: unknown): string | null {
  if (value === null || value === undefined) return null;
  const candidate = String(value).trim();
  return candidate && STABLE_ID_PATTERN.test(candidate) ? candidate : null;
}

export function isTurnScopedSettlementOutcome(
  deliveryOutcome: unknown,
  progressObservation: unknown,
  expectedSettlementBindingId: string | null,
): boolean {
  const normalizedOutcome = String(deliveryOutcome ?? "").trim() as DeliveryOutcome;
  if (PROGRESS_DELIVERY_OUTCOMES.has(normalizedOutcome)) return true;
  if (
    normalizedOutcome !== "outcome_gap" ||
    expectedSettlementBindingId === null
  ) {
    return false;
  }
  const observation = jsonObject(progressObservation);
  if (
    !observation ||
    observation.schema_version !== "typed_progress_observation_v0" ||
    observation.result_class !== "blocked" ||
    stableIdentifier(observation.work_item_id) !== expectedSettlementBindingId ||
    stableIdentifier(observation.blocker_id) === null ||
    !Array.isArray(observation.evidence_ids) ||
    observation.evidence_ids.length === 0
  ) {
    return false;
  }
  return observation.evidence_ids.every(
    (evidenceId) => stableIdentifier(evidenceId) !== null,
  );
}

export function decodeOptionalDeliveryOutcome(
  value: unknown,
  label = "delivery_outcome",
): DeliveryOutcome | null {
  if (value === null || value === undefined || value === "") return null;
  if (DELIVERY_OUTCOMES.some((candidate) => candidate === value)) {
    return value as DeliveryOutcome;
  }
  throw new EffectRuntimeRequestError(`${label} is unsupported`);
}

export function decodeOptionalMaterialDeliveryOutcome(
  value: unknown,
  label = "delivery_outcome",
): MaterialDeliveryOutcome | null {
  const outcome = decodeOptionalDeliveryOutcome(value, label);
  if (
    outcome === null ||
    MATERIAL_DELIVERY_OUTCOMES.some((candidate) => candidate === outcome)
  ) {
    return outcome as MaterialDeliveryOutcome | null;
  }
  throw new EffectRuntimeRequestError(`${label} is not a material delivery outcome`);
}

export function isMaterialDeliveryOutcome(
  value: DeliveryOutcome | null,
): value is MaterialDeliveryOutcome {
  return MATERIAL_DELIVERY_OUTCOMES.some((candidate) => candidate === value);
}
